### PR TITLE
Implement reproducible container build process

### PR DESCRIPTION
**Pull request by yakman2020**

### CODE REVIEW

File: Docker/Dockerfile.bootstrap [hunk 1]
```diff
@@ -0,0 +1,154 @@
+FROM scratch
+
+# The way to get a trusted build is to start with a trusted build. If we don't have one,
+# we need to use untrusted components to perform the build, then rebuild with the provisionally trusted components, 
+# then compare. If we get different contents, we don't know where the problem is, but we would know there is a problem.
+
+# this can't come from a URL because it won't get unpacked. 
+# We need a tight secure CDN to hold golden images. 
+ADD ubuntu-base-18.04.3-base-amd64.tar.gz  /
```
Comment (BRMcLaren) [inline Docker/Dockerfile.bootstrap:9]: Where is this coming from? 

How do we know it has not been altered and how can external users check the build process this image has gone through.

Edit: it comes from a script generated locally.

File: Docker/Dockerfile.bootstrap [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+ADD https://oedownload.blob.core.windows.net/oe-build/intel-sgx/libsgx-enclave-common-dev_2.7.100.4-bionic1_amd64.deb?st=2019-10-25T16%3A55%3A41Z&se=2020-10-26T16%3A55%3A00Z&sp=rl&sv=2018-03-28&sr=c&sig=nJWY1Q%2FXHkdKRBtfKdjHEbhDZU8zTOZ6Mc0JTjZK2gU%3D /tmp/libsgx-enclave-common-dev+2.7.100.4-bionic_amd64.deb
```
Comment (BRMcLaren) [inline Docker/Dockerfile.bootstrap:10]: How is this CDN secured and changes tracked?

Further, how are maintainers going to upload new drivers etc. in a secure way?

File: Docker/Dockerfile.bootstrap [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+ADD https://oedownload.blob.core.windows.net/oe-build/intel-sgx/libsgx-enclave-common_2.7.100.4-bionic1_amd64.deb?st=2019-10-25T16%3A55%3A41Z&se=2020-10-26T16%3A55%3A00Z&sp=rl&sv=2018-03-28&sr=c&sig=nJWY1Q%2FXHkdKRBtfKdjHEbhDZU8zTOZ6Mc0JTjZK2gU%3D /tmp/libsgx-enclave-common+2.7.100.4-bionic_amd64.deb
+#ADD https://oebuildinfo.blob.core.windows.net/buildinfo/ubuntu-base-18.04.3-base-amd64.tar.gz.sig
```
Comment (BRMcLaren) [inline Docker/Dockerfile.bootstrap:14]: This seems to be commented out?

File: Docker/Dockerfile.bootstrap [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+#
+# Check the signature.
+
+#
+# Rewrite sources.list to only get from CDN. We would like to trust ubu, but don't
+# 2Do: setup CDN and build dummy
```
Comment (BRMcLaren) [inline Docker/Dockerfile.bootstrap:21]: ToDo

File: Docker/Dockerfile.bootstrap [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+RUN DEBIAN_FRONTEND=noninteractive apt -y update --fix-missing
+RUN DEBIAN_FRONTEND=noninteractive apt -y dist-upgrade
+RUN DEBIAN_FRONTEND=noninteractive apt-get -y install musl musl-dev ninja-build python3 xz-utils libxml2 libgcc-7-dev make git zlib1g zlib1g-dev libreadline-dev libsqlite3-dev libbz2-dev libffi-dev liblzma-dev python-openssl gcc-7 g++-7 gpg libssl-dev dpkg-dev
+
+RUN /bin/bash -c "ln -sf /usr/bin/gcc-7 /usr/bin/gcc ; ln -sf /usr/bin/g++-7 /usr/bin/g++"
+
+# Now cmake. 
+# 2do: Cmake should be coming from our CDN. Even if using the debian package was a good idea, the
+# cmake package version is 3.10.  We require 3.13 minimum.
+ADD https://github.com/Kitware/CMake/releases/download/v3.15.1/cmake-3.15.1.tar.gz  /tmp/cmake-3.15.1.tar.gz
```
Comment (BRMcLaren) [inline Docker/Dockerfile.bootstrap:47]: I was under the impression that Github was not a secure source, it uses SHA-1 and we probably do not want to use this as a source.

File: Docker/Dockerfile.bootstrap [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+#RUN /bin/bash -c "\
+#    pushd . ;\
+#    cd cmake-3.15.1-Linux-x86_64 ; \
+#    cp -Rf * /usr/local/ ; \
+#    popd \
+#    "
+ENV PATH=/usr/local/bin:/bin:/sbin:/usr/bin
+
+#
+# 2do: Get this file from CDN and check signature
```
Comment (BRMcLaren) [inline Docker/Dockerfile.bootstrap:68]: TODO

**Review by BRMcLaren — COMMENTED** [review]
There is a lot of commented out code, making this difficult to review and follow.

File: Docker/Dockerfile.bootstrap [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+#
+COPY build_toolchain.sh /tmp
+COPY signatures/ /tmp
+
+RUN rm -rf /usr/local ; mkdir -p /usr/local
+
+#
+#Unpack the golden cmake
+WORKDIR /tmp
+#RUN tar xvfz cmake-3.15.1-Linux-x86_64.tar.gz
```
Comment (BRMcLaren) [inline Docker/Dockerfile.bootstrap:58]: This seems to be commented out, what is the impact of including it?

File: Docker/Dockerfile.bootstrap [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+#
+# We bootstrap with a prebuilt compiler, but then build the toolchain and check the signatures against the 
+# prebuilt. If the signatures don't match we have a security issue.
+# 
+#ADD http://releases.llvm.org/7.1.0/llvm-7.1.0.src.tar.xz /tmp
+#ADD http://releases.llvm.org/7.1.0/llvm-7.1.0.src.tar.xz.sig /tmp
+ADD  https://oedownload.blob.core.windows.net/oe-build/llvm-src/llvm-7.1.0.src.tar.xz?st=2019-11-01T16%3A12%3A28Z&se=2020-11-02T16%3A12%3A00Z&sp=rl&sv=2018-03-28&sr=c&sig=qHu9rtuvCR6f5d9esJ7w6t7aaEp%2F4gyE5hmDd8uM0d8%3D /tmp/llvm-7.1.0.src.tar.xz
+ADD  https://oeverify.blob.core.windows.net/oe-build/llvm-src/llvm-7.1.0.src.tar.xz.sig?st=2019-11-01T16%3A09%3A42Z&se=2020-11-02T16%3A09%3A00Z&sp=rl&sv=2018-03-28&sr=c&sig=fL4Q3n0VIicenrz2yxm%2FWsS2qHBwWxm%2BeF8eiqOnAA0%3D /tmp/llvm-7.1.0.src.tar.xz.sig
+
+#RUN tar xvfJ /tmp/clang+llvm-7.0.1-x86_64-linux-gnu-ubuntu-18.04.tar.xz
```
Comment (BRMcLaren) [inline Docker/Dockerfile.bootstrap:123]: There is a lot of commented out code here, does this need to be included?

File: Docker/Dockerfile.bootstrap [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+#    cp -Rf * /usr/local/ ; \
+#    ldconfig /usr/local/lib ; \
+#    popd \
+#    "
+RUN mkdir -p /src
+RUN mkdir -p /build
+
+WORKDIR /src
+
+# Check the signature, build then check the sha256 hash of the binaries
```
Comment (BRMcLaren) [inline Docker/Dockerfile.bootstrap:137]: Checking the signature?

**Comment by BRMcLaren** [comment]
I am having trouble following what steps need to be run. After trying to create a bootstrap image with build_bootstrap.sh and then try to build the toolchain...

`brmclare@cdpxTest:~/openenclave-build/Docker$ ./build_toolchain.sh 
gpg: can't open '/tmp/tstellar-gpg-key.asc': No such file or directory
gpg: Total number processed: 0
could not verify signature  /tmp/cfe-7.1.0.src.tar.xz.sig
~/openenclave-build/Docker ~/openenclave-build/Docker
rm: cannot remove '/src/cmake-3.15.1/CMakeCache.txt': No such file or directory
rm: cannot remove '/build/cmake-3.15.1/CMakeCache.txt': No such file or directory`

Which then deleted everything within the Docker file.

I spent a few hours code reviewing this and clarification is needed.

Can you please add the rest of the documentation?
Can you please add the design documentation, as this needs to be reviewed?
Can you clarify why there is a Docker folder which includes everything, is this intended?
Please, include the DockerFile that runs this step by step to generate the reproducible container.
Please, update the code to not include commented out code sections as this affects readability.
Please, clarify how we get the code sources into the blobs in a secure manner and how this is double-checked against the hashes when downloaded?

We may need to break this up over several pull requests as well, a 1,649 line change is too large.

**Review by BRMcLaren — CHANGES_REQUESTED** [review]

File: Docker/Dockerfile.bootstrap (hunk 1 above)
Comment (yakman2020) [reply Docker/Dockerfile.bootstrap:9]: Correct. It gets downloaded by the script "build_bootstrap.sh" from our CDN.  The reason a script is needed is that we don't trust dockerhub.

File: Docker/Dockerfile.bootstrap (hunk 2 above)
Comment (yakman2020) [reply Docker/Dockerfile.bootstrap:10]: Access is controlled by an SAS.  The write access is currently only me, but we can add a role for updateds via the portal.

File: Docker/Dockerfile.bootstrap (hunk 5 above)
Comment (yakman2020) [reply Docker/Dockerfile.bootstrap:47]: We don't like depending on github. This should be a tar file from the CDN.

**Pull request closed without merging**